tutorials/envoy-flask-google-container-engine.md

```diff
@@ -0,0 +1,566 @@
+
+A useful pattern is to enlist a proxy, like [Envoy](https://lyft.github.io/envoy/) to help [make your application more resilient and observable](https://www.datawire.io/guide/traffic/getting-started-lyft-envoy-microservices-resilience/). Envoy can be a bit daunting to set up, so this tutorial will walk you through deploying a Python Flask webapp with Envoy on Google Container Engine.
+
+## The Application
+
+Our application is a super simple REST-based user service: it can create users, read information about a user, and process simple logins. Even a trivial application like this involves several real-world concerns, though:
+
+* It requires persistent storage.
+* It will let us explore scaling the different pieces of the application.
+* It will let us explore Envoy at the edge, where the user’s client talks to our application.
+* It will let us explore Envoy internally, brokering communications between the various parts of the application.
+
+Envoy runs as a sidecar, so it's language-agnostic. For this tutorial, we’ll use Python and Flask for simplicity. We’ll use PostgreSQL for persistence because it has good Python support and it’s easy to get running both locally and in the cloud. And of course, running on Google Container Engine means managing everything with Kubernetes.
+
+## Setting Up
+
+### Google Container Engine
+
+You'll need a Google Cloud Platform account for this, so that you can set up a Google Container Engine cluster, just visit the [Google Cloud Platform Console](https://console.cloud.google.com/kubernetes) and use the UI to create a new cluster. Picking the defaults should be fine for our purposes.
+
+### Kubernetes
+
+You'll need `kubectl`, the Kubernetes CLI, to work with Google Container Engine. On a Mac you can use `brew install kubernetes-cli`; else you'll need to follow the [Kubernetes installation intructions](https://kubernetes.io/docs/user-guide/prereqs/).
+
+### Docker
+
+Google Compute Engine runs code from Docker images, so you'll need the Docker CLI, `docker`. [Docker Community Edition](https://www.docker.com/community-edition) is fine if you're just getting started (again, on a Mac, the easy way is `brew install docker`).
+
+### The Application
+
+Everything we'll use in this demo is in [Datawire's `envoy-steps` repo](https://github.com/datawire/envoy-steps), so you'll need to clone that and `cd` into your clone:
+
+```
+git clone https://github.com/datawire/envoy-steps.git
+cd envoy-steps
+```
+
+In your `envoy-steps` directory, you should see `README.md` and directories named `postgres`, `usersvc`, etc. Each of those directories contains a service to be deployed, and each can be brought up or down independently with
+
+```
+sh up.sh $service
```

Comment:
That's basically paranoia on my part. `./up.sh` requires correct permissions and a correct shebang line. `sh up.sh` requires neither of those things, so I tend to gravitate to that form for installation instructions (even though both _should_ be correct here......).